File: python/terminal.py
import random


class Generation:
    def __init__(self, width: int, height: int):
        self.cells = [[0 for _ in range(width)] for _ in range(height)]

        self.width = width
        self.height = height

    def display(self):
        for y in range(self.height):
            for x in range(self.width):
                print("●" if self.cells[y][x] else " ", end="")

            print()

    def random(self):
        for y in range(self.height):
            for x in range(self.width):
                if random.random() > 0.90:
                    self.cells[y][x] = 1

    def calculate_alive_neigbours(self, ix: int, iy: int):
        amount = 0

        for dy in range(-1, 2):
            for dx in range(-1, 2):
                if dx == 0 and dy == 0:
                    continue

                x = (ix + dx) % self.width
                y = (iy + dy) % self.height

                if self.cells[y][x]:
                    amount += 1

        return amount

    def update(self):
        next_gen = self.__class__(self.width, self.height)

        for y in range(self.height):
            for x in range(self.width):
                alive_neighbours = self.calculate_alive_neigbours(x, y)

                current_cell = self.cells[y][x]

                if (
                    current_cell and (alive_neighbours == 2 or alive_neighbours == 3)
                ) or (not current_cell and alive_neighbours == 3):
                    next_state = 1
                else:
                    next_state = 0

                next_gen.cells[y][x] = next_state

        self.cells = next_gen.cells


def play():
    generation = Generation(170, 170)
    generation.random()

    while True:
        generation.display()
        generation.update()


play()

File: javascript/terminal.js
function mod(a, b) {
    let r = a % b;
    return r < 0 ? r + b : r;
}

class Generation {
    constructor(width, height) {
        this.cells = new Array(height);

        for (let y = 0; y < height; y++) {
            this.cells[y] = new Array(width).fill(0);
        }

        this.width = width;
        this.height = height;
    }

    random() {
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                this.cells[y][x] = Math.floor(Math.random() * 2);
            }
        }
    }

    display() {
        for (let y = 0; y < this.height; y++) {
            let line = "";

            for (let x = 0; x < this.width; x++) {
                line += this.cells[y][x] ? "●" : " ";
            }

            console.log(line);
        }
    }

    calculate_alive_neigbours(ix, iy) {
        let amount = 0;

        for (let dy = -1; dy < 2; dy++) {
            for (let dx = -1; dx < 2; dx++) {
                if (dx == 0 && dy == 0) {
                    continue;
                }

                let x = mod(ix + dx, this.width);
                let y = mod(iy + dy, this.height);

                if (this.cells[y][x]) {
                    amount += 1;
                }
            }
        }

        return amount;
    }

    update() {
        let next_gen = new Generation(this.width, this.height);

        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                let alive_neighbours = this.calculate_alive_neigbours(x, y);

                let current_cell = this.cells[y][x];

                let next_state =
                    (current_cell &&
                        (alive_neighbours == 2 || alive_neighbours == 3)) ||
                    (!current_cell && alive_neighbours == 3)
                        ? 1
                        : 0;

                next_gen.cells[y][x] = next_state;
            }
        }

        this.cells = next_gen.cells;
    }
}

let generation = new Generation(170, 170);

function play() {
    generation.random();

    while (true) {
        generation.display();
        generation.update();
    }
}

play();
